### PR TITLE
VIDEO-3560 - addProcessor API

### PR DESCRIPTION
**Pull request by charliesantos**
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

This PR:

Implements addProcessor API base on the following specifications:
- IDL: https://code.hq.twilio.com/client/video-sdk-api/pull/77
- FRD: https://code.hq.twilio.com/client/sdk-frd/pull/63

TODO (in the removeProcessor ticket):

- Implement removeProcessor
- Account for track.restart and track.stop

### CODE REVIEW

**Review by manjeshbhargav — CHANGES_REQUESTED** [review]
@charliesantos Great work! One suggestion: we need to make sure that if `.mediaStreamTrack` is one of disabled, muted or ended, then we need to stop capturing. And, if it later gets enabled or unmuted, then we need to resume capturing.

File: lib/media/track/localvideotrack.js [hunk 1]
```diff
@@ -69,12 +69,31 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
     return super._end.apply(this, arguments);
   }
 
+  addProcessor() {
```
[the remark this answers is not in the document]
Comment (charliesantos) [reply lib/media/track/localvideotrack.js:72]: The documentation is on VideoTrack.js, which is inherited here.

**Comment by charliesantos** [comment]
>One suggestion: we need to make sure that if .mediaStreamTrack is one of disabled, muted or ended, then we need to stop capturing. And, if it later gets enabled or unmuted, then we need to resume capturing.

Good one @manjeshbhargav . I''m planning on doing this in the removeProcessor ticket. Basically anything that requires stopping and removing the the capture process, I'll take care of it in that ticket.

File: lib/media/track/localvideotrack.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/media/track/localvideotrack.js:72]: @charliesantos But the context is different between LocalVideoTrack and RemoteVideoTrack. You can give different examples for each.

[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/media/track/localvideotrack.js:72]: @charliesantos But the context is different between LocalVideoTrack and RemoteVideoTrack. You can give different examples for each.

File: lib/media/track/localvideotrack.js [hunk 2]
```diff
@@ -69,12 +69,31 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
...
+
+    return result;
+  }
+
   /**
    * Disable the {@link LocalVideoTrack}. This is effectively "pause".
    * @returns {this}
    * @fires VideoTrack#disabled
    */
   disable() {
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/media/track/localvideotrack.js:94]: I think you can reword it to:
```
Disable the {@link LocalVideoTrack}. This is effectively "pause". If a {@link VideoProcessor} is added, then
`processedTrack` is disabled as well. 
```

File: lib/media/track/localvideotrack.js [hunk 3]
```diff
@@ -92,6 +111,9 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
    * @fires VideoTrack#enabled
    */
   enable() {
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply lib/media/track/localvideotrack.js:115]: I think this can be reworded to:
```
Enable the {@link LocalVideoTrack}. This is effectively "unpause". If a {@link VideoProcessor} is added, then
`processedTrack` is enabled as well. 
```

**Review by manjeshbhargav — COMMENTED** [review]
@charliesantos I recommend that you address the stopping of capturing upon the track being ended/muted/disabled in this PR since we may need to provide this to the customer before implementing `removeProcessor()`. I'm worried about the performance implications when a track is disabled and a VideoProcessor is still running.

File: package.json [hunk 4]
```diff
@@ -91,7 +92,7 @@
     "test:unit": "mocha ./test/unit/index.js",
     "test:integration:adapter": "node ./scripts/karma.js karma/integration.adapter.conf.js",
     "test:integration": "node ./scripts/karma.js karma/integration.conf.js",
-    "test:umd:install": "npm install puppeteer",
+    "test:umd:install": "npm install puppeteer@5.5.0",
```
Comment (charliesantos) [inline package.json:95]: @manjeshbhargav @makarandp0 @PikaJoyce I added this version because puppeteer released a major version with breaking changes recently which broke our tests. I'm locking the version so it doesn't auto update. Let me know if you have any concerns.

File: lib/media/track/localvideotrack.js [hunk 5]
```diff
@@ -69,6 +69,30 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
     return super._end.apply(this, arguments);
   }
 
+  /**
+   * Add a {@link VideoProcessor} to allow for custom processing of video frames belonging to a VideoTrack.
+   * Only Chrome supports this as of now.
+   * @param {VideoProcessor} processor - The {@link VideoProcessor} to use.
+   * @returns {this}
+   * @example
+   * class Processor {
```
Comment (manjeshbhargav) [inline lib/media/track/localvideotrack.js:78]:
```suggestion
   * class GrayScaleProcessor {
```

File: lib/media/track/localvideotrack.js [hunk 6]
```diff
@@ -69,6 +69,30 @@ class LocalVideoTrack extends LocalMediaVideoTrack {
...
+   *
+   *     const context = this.outputFrame.getContext('2d');
+   *     context.filter = 'grayscale(100%)';
+   *     context.drawImage(inputFrame, 0, 0, inputFrame.width, inputFrame.height);
+   *     return this.outputFrame;
+   *   }
+   * }
+   *
+   * const localVideoTrack = Array.from(room.localParticipant.videoTracks.values())[0].track;
+   * localVideoTrack.addProcessor(new Processor());
```
Comment (manjeshbhargav) [inline lib/media/track/localvideotrack.js:94]:
```suggestion
   * localVideoTrack.addProcessor(new GrayScaleProcessor());
```

File: lib/media/track/videotrack.js [hunk 7]
```diff
@@ -104,7 +104,12 @@ class VideoTrack extends MediaTrack {
                 .drawImage(outputFrame, 0, 0, this._outputFrame.width, this._outputFrame.height);
             }
           })
-          .finally(() => captureFrame(process));
+          .finally(() => {
+            const { enabled, muted, readyState } = this.mediaStreamTrack;
+            if (enabled && !muted && readyState === 'live') {
+              captureFrame(process);
+            }
```
Comment (manjeshbhargav) [inline lib/media/track/videotrack.js:111]: I think this should be checked before processing the current frame instead of scheduling the next frame. WDYT?

File: test/unit/spec/media/track/videotrack.js [hunk 8]
```diff
@@ -124,9 +124,17 @@ describe('VideoTrack', () => {
           sinon.assert.calledOnce(videoTrack._updateElementsMediaStreamTrack);
         });
 
-        it('should start capturing frames', () => {
+        it('should start capturing frames after setting the processor', () => {
           sinon.assert.calledOnce(videoTrack._captureFrames);
         });
+
+        it('should start capturing frames whenever mediaStreamTrack unmutes', () => {
```
Comment (manjeshbhargav) [inline test/unit/spec/media/track/videotrack.js:131]: What about tests for `enabled`?

**Review by manjeshbhargav — APPROVED** [review]
@charliesantos LGTM.

**Review by PikaJoyce — APPROVED** [review]
Great job Charlie! Very cool!

File: test/unit/spec/media/track/videotrack.js (hunk 8 above)
Comment (charliesantos) [reply test/unit/spec/media/track/videotrack.js:131]: Since `.enable` already calls captureFrames, we already have that test [here](https://github.com/twilio/twilio-video.js/pull/1374/files#diff-7925908b69cda71b02214829e73cb7b620e9515975759a6e44e96fd3c8c2dea3R90).
